vrx_gz/src/WildlifeScoringPlugin.cc

```diff
@@ -0,0 +1,810 @@
+
+  /// \brief The goal is to circumnavigate the buoy clockwise.
+  CIRCUMNAVIGATE_CLOCKWISE,
+
+  /// \brief The goal is to circumnavigate the buoy counterclockwise
+  CIRCUMNAVIGATE_COUNTERCLOCKWISE,
+};
+
+/// \brief All buoy states.
+enum class BuoyState
+{
+  /// \brief Not "in" the gate and never engaged.
+  NEVER_ENGAGED,
+
+  /// \brief Not "in" the gate but was engaged at some point.
+  NOT_ENGAGED,
+
+  /// \brief Inside the area of activation of the gate.
+  ENGAGED,
+
+  /// \brief Succesfully circumnavigated.
+  CIRCUMNAVIGATED,
+};
+
+/// \brief All gate states.
+enum class GateState
+{
+  /// \brief Not "in" the gate.
+  VEHICLE_OUTSIDE,
+
+  /// \brief Before the gate.
+  VEHICLE_BEFORE,
+
+  /// \brief After the gate.
+  VEHICLE_AFTER,
+
+  /// \brief Gate crossed!
+  CROSSED,
+};
+
+/// \brief A virtual gate to help detecting circumnavigation.
+class VirtualGate
+{
+  /// \brief Constructor.
+  /// \param[in] _leftMarkerEntity The Entity of the buoy.
+  /// \param[in] _offset The offset from the buoy that delimitates the gate.
+  /// \param[in] _width The width of the gate.
+  /// \param[in out] _ecm The Entity Component Manager.
+  public: VirtualGate(const sim::Entity _leftMarkerEntity,
+                      const math::Vector3d &_offset,
+                      double _width,
+                      sim::EntityComponentManager &_ecm);
+
+  /// \brief Where is the given robot pose with respect to the gate?
+  /// \param _robotWorldPose Pose of the robot, in the world frame.
+  /// \return The gate state given the current robot pose.
+  public: GateState IsPoseInGate(const math::Pose3d &_robotWorldPose) const;
+
+  /// \brief Recalculate the pose of the gate.
+  /// \param[in] _ecm The Entity Component Manager.
+  public: void Update();
+
+  /// \brief The left marker (buoy) entity.
+  public: sim::Entity leftMarkerEntity = sim::kNullEntity;
+
+  /// \brief The offset of the right marker with respect to the left marker
+  /// in world coordinates.
+  public: math::Vector3d offset;
+
+  /// \brief The width of the gate in meters.
+  public: double width;
+
+  /// \brief The center of the gate in the world frame. Note that the roll and
+  /// pitch are ignored. Only yaw is relevant and it points into the direction
+  /// in which the gate should be crossed.
+  public: math::Pose3d pose;
+
+  /// \brief The state of this gate.
+  public: GateState state = GateState::VEHICLE_OUTSIDE;
+
+  /// \brief The Entity Component Manager.
+  public: sim::EntityComponentManager &ecm;
+};
+
+/////////////////////////////////////////////////
+VirtualGate::VirtualGate(const sim::Entity _leftMarkerEntity,
+  const math::Vector3d &_offset, double _width,
+  sim::EntityComponentManager &_ecm)
+  : leftMarkerEntity(_leftMarkerEntity),
+    offset(_offset),
+    width(_width),
+    ecm(_ecm)
+{
+  this->Update();
+}
+
+/////////////////////////////////////////////////
+void VirtualGate::Update()
+{
+  if (this->leftMarkerEntity == sim::kNullEntity)
```

Comment:
I don't think so because we are also checking in the next line if the component was found. Removed in [2f77bdf](https://github.com/osrf/vrx/pull/547/commits/2f77bdf922221ca204aa17714c9f7d4bb021b1fd).